src/crystal/hasher.cr

```diff
@@ -6,15 +6,68 @@ struct Crystal::Hasher
   # value for primitive and basic Crystal objects. All other
   # hashes are computed based on these.
   #
-  # TODO: the implementation is naive and should probably use
-  # another algorithm like SipHash or FNV.
+  # The algorithm bases on https://github.com/funny-falcon/funny_hash
+  #
+  # It is two multiply-rotate 64bit hash functions, combined
+  # within finalizer.
+  #
+  # Both hash functions combines previous state with a block value
+  # before multiplication. One function multiplies new state
+  # as is (and then rotates state), other multiplies new state
+  # already rotated by 32 bits.
+  #
+  # This way algorithm ensures that every block bit affects at
+  # least 1 bit of every state, and certainly many bits of some
+  # state. So effect of this bit could not be easily canceled
+  # with following blocks. (Cause next blocks have to cancel
+  # bits on non-intersecting positions in both states).
+  # Rotation by 32bit with multiplication also provides good
+  # inter-block avalanche.
+  #
+  # Finalizer performs murmur-like finalization on both functions,
+  # and then combines them with addition. It greatly reduce
+  # possibility of state deduction.
+  #
+  # Note, it provides good protection from HashDos if and only if:
+  # - seed is securely random and not exposed to attacker,
+  # - hash result is also not exposed to attacker in a way other
+  #   than effect of using it in Hash implementation.
+  # Do not output calculated hash value to user's console/form/
+  # html/api response, etc. Use some from digest package instead.
+
+  @@seed = uninitialized UInt64[2]
+  Random::Secure.random_bytes(Slice.new(pointerof(@@seed).as(UInt8*), sizeof(typeof(@@seed))))
+
+  def initialize(@a : UInt64 = @@seed[0], @b : UInt64 = @@seed[1])
+  end
+
+  private C1 = 0xacd5ad43274593b9_u64
+  private C2 = 0x6956abd6ed268a3d_u64
 
-  @@seed = uninitialized UInt64
-  Random::Secure.random_bytes(Slice.new(pointerof(@@seed).as(UInt8*), 8))
+  private def rotl32(v : UInt64)
+    (v << 32) | (v >> 32)
+  end
 
-  property result : UInt64 = @@seed
+  private def permute(v : UInt64)
+    @a = rotl32(@a ^ v) * C1
+    @b = (rotl32(@b) ^ v) * C2
+    self
+  end
+
+  def result
+    a, b = @a, @b
+    a ^= (a >> 23) ^ (a >> 40)
+    b ^= (b >> 23) ^ (b >> 40)
+    a *= C1
+    b *= C2
+    a ^= a >> 32
+    b ^= b >> 32
+    a + b
+  end
 
   def nil
+    @a += @b
+    @b += 1
```

Comment:
Earlier you use LFSR here, it is not wanted anymore?

Reply:
Both state variable should change in some way.

Combination of addition on @a and lsfr on @b gives period almost `2**128`. That is great, but unnecessary.

Current version is "quadrating probing", and gives period `2**64`. I think, it is just enough.

Even `2**32` could be enough for intended use case, because if someone triggers `2**32` calls to `nil.hash(hasher)`,

Reply:
if some triggers `2**32` calls to `nil.hash(hasher)`,  they are already successful at DoS-ing.